File: frontend/lib/sse.ts
export type SSEEvent = { event: string; data: unknown };

// Incremental SSE parser. Fetch chunks do not line up with event boundaries, so
// leftover text is kept between feed() calls.
export class SSEParser {
  private buffer = "";

  feed(chunk: string): SSEEvent[] {
    this.buffer += chunk.replace(/\r\n/g, "\n");
    const events: SSEEvent[] = [];
    let separator = this.buffer.indexOf("\n\n");
    while (separator !== -1) {
      const block = this.buffer.slice(0, separator);
      this.buffer = this.buffer.slice(separator + 2);
      const parsed = parseBlock(block);
      if (parsed) events.push(parsed);
      separator = this.buffer.indexOf("\n\n");
    }
    return events;
  }
}

function parseBlock(block: string): SSEEvent | null {
  let event = "message";
  const dataLines: string[] = [];
  for (const line of block.split("\n")) {
    if (!line || line.startsWith(":")) continue;
    const colon = line.indexOf(":");
    const field = colon === -1 ? line : line.slice(0, colon);
    let value = colon === -1 ? "" : line.slice(colon + 1);
    if (value.startsWith(" ")) value = value.slice(1);
    if (field === "event") event = value;
    else if (field === "data") dataLines.push(value);
  }
  if (dataLines.length === 0) return null;
  return { event, data: JSON.parse(dataLines.join("\n")) };
}

File: henet_kb/api/sse.py
import json
from typing import Any


def encode_event(event: str, data: dict[str, Any]) -> str:
    """One SSE event. Data is always a single JSON line."""
    return f"event: {event}\ndata: {json.dumps(data, ensure_ascii=False)}\n\n"


class SSEParser:
    """Incremental SSE parser. Handles events split across chunks."""

    def __init__(self) -> None:
        self._buffer = ""

    def feed(self, chunk: str) -> list[tuple[str, dict[str, Any]]]:
        self._buffer += chunk.replace("\r\n", "\n")
        events: list[tuple[str, dict[str, Any]]] = []
        while "\n\n" in self._buffer:
            block, self._buffer = self._buffer.split("\n\n", 1)
            parsed = self._parse_block(block)
            if parsed is not None:
                events.append(parsed)
        return events

    @staticmethod
    def _parse_block(block: str) -> tuple[str, dict[str, Any]] | None:
        event = "message"
        data_lines: list[str] = []
        for line in block.split("\n"):
            if not line or line.startswith(":"):
                continue
            field, _, value = line.partition(":")
            value = value[1:] if value.startswith(" ") else value
            if field == "event":
                event = value
            elif field == "data":
                data_lines.append(value)
        if not data_lines:
            return None
        return event, json.loads("\n".join(data_lines))
